### PR TITLE
Implement the voting system 

### PR DESCRIPTION
**Pull request by MSNTCS**

### CODE REVIEW

File: votingsystem/src/common.rs [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+impl fmt::Display for NetworkId {
+    fn fmt(&self, f: &mut fmt::Formatter) -> Result<(), fmt::Error> {
+        let s = std::str::from_utf8(&self.0).expect("network_id a valid utf8 string");
+        write!(f, "{}", s)
+    }
+}
+
+impl Default for NetworkId {
+    fn default() -> Self {
+        NetworkId([116, 99])
```
Comment (majecty) [inline votingsystem/src/common.rs:35]: Let's use character literal.
```suggestion
        NetworkId([b't', b'c'])
```

I chanted the same code in timestamp app: https://github.com/CodeChain-io/foundry/pull/453/files

File: votingsystem/src/general_meeting.rs [hunk 2]
```diff
@@ -0,0 +1,262 @@
...
+    fn execute_transaction(&mut self, transaction: &Transaction) -> Result<TransactionOutcome, ()> {
+        if let Err(error) = self.excute_tx(transaction) {
+            match error {
+                ExecuteError::InvalidMetadata => Err(()),
+                ExecuteError::InvalidSign => Err(()),
+                ExecuteError::InvalidFormat => Err(()),
+            }
+        } else {
+            Ok(Default::default())
+        }
```
Comment (majecty) [inline votingsystem/src/general_meeting.rs:117]: Why didn't you call `execute_tx` in this function?

File: votingsystem/src/general_meeting.rs [hunk 3]
```diff
@@ -0,0 +1,262 @@
...
+
+    fn block_closed(&mut self) -> Result<Vec<Event>, CloseBlockError> {
+        Ok(Vec::new())
+    }
+}
+
+#[derive(Debug, Serialize, Deserialize, Clone)]
+pub struct TimeStamp{
+    time: u64,
+}
```
Comment (majecty) [inline votingsystem/src/general_meeting.rs:136]: It will be serialized as `{ time: 3 }` instead of `3`.
Is it your intention?

File: votingsystem/src/general_meeting.rs [hunk 4]
```diff
@@ -0,0 +1,262 @@
...
+impl TimeStamp {
+    fn set_time(&mut self, time: u64){
+        self.time = time;
+    }
+}
+
+#[derive(Debug, Serialize, Deserialize, Clone)]
+pub struct GeneralMeetingId {
+    id: u8,
+}
```
Comment (majecty) [inline votingsystem/src/general_meeting.rs:148]: It will be serialized as `{ id: 3 }` instead of `3`.
Is it your intention?

File: votingsystem/src/general_meeting.rs [hunk 5]
```diff
@@ -0,0 +1,262 @@
...
+#[derive(Debug, Serialize, Deserialize)]
+pub enum GeneralMeetingError {
+    NoSuchMeeting,
+    MeetingExist,
+}
+
+#[derive(Debug, Serialize, Deserialize, Clone)]
+pub struct GeneralMeeting {
+    id: GeneralMeetingId,
+    number_of_options: u32,
```
Comment (majecty) [inline votingsystem/src/general_meeting.rs:177]: Do we need `number_of_options`?

File: votingsystem/src/general_meeting.rs [hunk 6]
```diff
@@ -0,0 +1,262 @@
...
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+pub enum GeneralMeetingError {
+    NoSuchMeeting,
+    MeetingExist,
+}
+
+#[derive(Debug, Serialize, Deserialize, Clone)]
+pub struct GeneralMeeting {
```
Comment (majecty) [inline votingsystem/src/general_meeting.rs:175]: Where is `number_of_agendas`?

File: votingsystem/src/general_meeting.rs [hunk 7]
```diff
@@ -0,0 +1,262 @@
...
+    voting_end_time: TimeStamp,
+    tallying_time: TimeStamp,
+    result: Option<Vec<VoteResult>>,
+}
+
+impl GeneralMeeting {
+    pub fn new(end_time: TimeStamp, tallying_time: TimeStamp) -> Self {
+        let meeting_id= GeneralMeetingId::new();
+        let options = 4;
+        let final_result = Some(Vec::new());
```
Comment (majecty) [inline votingsystem/src/general_meeting.rs:187]: If there is no final_result, isn't using `None` better?

File: votingsystem/src/general_meeting.rs [hunk 8]
```diff
@@ -0,0 +1,262 @@
...
+}
+
+impl InitGenesis for Context {
+    fn begin_genesis(&mut self) {}
+
+    fn init_genesis(&mut self, config: &[u8]) {
+    }
+
+    fn end_genesis(&mut self) {}
+}
```
Comment (majecty) [inline votingsystem/src/general_meeting.rs:100]: If you don't do anything in functions in InitGenesis, why do you implement it?

File: votingsystem/src/general_meeting.rs [hunk 9]
```diff
@@ -0,0 +1,262 @@
...
+        }
+    }
+
+    fn check_transaction(&self, transaction: &Transaction) -> Result<(), coordinator::types::ErrorCode> {
+        let todo_fixthis: coordinator::types::ErrorCode = 3;
+        assert_eq!(transaction.tx_type(), "CreateGeneralMeeting");
+        let tx: OwnTransaction = serde_cbor::from_slice(&transaction.body()).map_err(|_| todo_fixthis)?;
+        tx.verify().map_err(|_| todo_fixthis)?;
+        Ok(())
+    }
```
Comment (majecty) [inline votingsystem/src/general_meeting.rs:126]: Shouldn't we check the signer?

File: votingsystem/src/general_meeting.rs [hunk 10]
```diff
@@ -0,0 +1,262 @@
...
+
+enum ExecuteError {
+    InvalidMetadata,
+    InvalidSign,
+    InvalidFormat,
+}
+
+struct Context{
+    pub storage: Option<Box<dyn SubStorageAccess>>,
+    pub meeting: Option<Box<dyn MeetingManager>>,
```
Comment (majecty) [inline votingsystem/src/general_meeting.rs:39]: Why did you add `meeting` in Context?
It is not a remote service.

File: votingsystem/src/general_meeting.rs [hunk 11]
```diff
@@ -0,0 +1,262 @@
...
+            tallying_time,
+            result: final_result
+        }
+    }
+}
+
+#[service]
+pub trait MeetingManager: Service {
+    fn create_general_meeting(&mut self, general_meeting: GeneralMeeting, id: GeneralMeetingId) -> Result<GeneralMeetingId,GeneralMeetingError>;
+}
```
Comment (majecty) [inline votingsystem/src/general_meeting.rs:201]: Why did you create `MeetingManager` service?
Which module will use this service?

File: votingsystem/src/general_meeting.rs (hunk 8 above)
Comment (MSNTCS) [reply votingsystem/src/general_meeting.rs:100]: Here is the part that is not complete and needs our discussion because I do not have a good insight of InitGenesis.

File: votingsystem/src/general_meeting.rs (hunk 9 above)
Comment (MSNTCS) [reply votingsystem/src/general_meeting.rs:126]: A verified person must be allowed to create a general meeting.

File: votingsystem/src/general_meeting.rs [hunk 12]
```diff
@@ -0,0 +1,271 @@
...
+            id: random_id
+        }
+    }
+}
+
+impl AsRef<[u8]> for GeneralMeetingId{
+    fn as_ref(&self)->&[u8]{
+        self.as_ref()
+    }
+}
```
Comment (majecty) [inline votingsystem/src/general_meeting.rs:171]: Did you add `as_ref` for the serialization? I think AsRef is not related to the serialization.
You can use `#[serde(transparent)]` to serialize a struct that has only a field exactly the same as if its one field were serialized itself.

https://serde.rs/container-attrs.html#transparent

I recommend you to read the serde document. It isn't long.

File: votingsystem/src/general_meeting.rs [hunk 13]
```diff
@@ -0,0 +1,271 @@
...
+enum ExecuteError {
+    InvalidMetadata,
+    InvalidSign,
+    InvalidFormat,
+    InvalidAdmin,
+}
+
+struct Context{
+    pub storage: Option<Box<dyn SubStorageAccess>>,
+    pub admin: Option<Public>,
```
Comment (majecty) [inline votingsystem/src/general_meeting.rs:42]: How about reading `admin` in the BlockChain state always?
If the admin value in the state changes, the GeneralMeeting module must read the value.
Currently, we are not changing it, but I predict it will be changed.

File: votingsystem/src/general_meeting.rs [hunk 14]
```diff
@@ -0,0 +1,271 @@
...
+    result: Option<Vec<VoteResult>>,
+}
+
+impl GeneralMeeting {
+    pub fn new(end_time: TimeStamp, tallying_time: TimeStamp) -> Self {
+        let meeting_id= GeneralMeetingId::new();
+        let options = 4;
+        Self{
+            id: meeting_id,
+            number_of_agends: options,
```
Comment (majecty) [inline votingsystem/src/general_meeting.rs:213]: I used the term "agenda" as the number of votes in the meeting.
For example, if there are two agendas in the meeting like "Do you like Kimchi" and "Do you know Bibimbap", the voters should vote two times like "I don't like kimchi" and "I don't know Bibimbap".

So the number of agenda cannot be hard-coded. Someone who creates a general meeting could set the number of agendas in the meeting.

File: votingsystem/src/general_meeting.rs (hunk 12 above)
Comment (majecty) [reply votingsystem/src/general_meeting.rs:171]: OK, I understand your intention.

I thought you used as_ref for the value serialization.
How about naming the [u8] as `Key` to express your intention?

```rust
type StorageKeyRef<'a> = &'a [u8];

impl AsRef<StorageKey> for GeneralMeetingId{
    fn as_ref(&self)->StorageKeyRef {
      std::slice::from_ref(&self.id)
    }
}
```

Also calling `self.as_ref()` in the as_ref definition is a infinite loop. If you want to make &T to &[T] you should use `std::slice::from_ref`: https://doc.rust-lang.org/std/slice/fn.from_ref.html

File: votingsystem/src/general_meeting.rs (hunk 3 above)
Comment (majecty) [reply votingsystem/src/general_meeting.rs]: @MSNTCS Could you check this comment?

File: votingsystem/src/general_meeting.rs (hunk 4 above)
Comment (majecty) [reply votingsystem/src/general_meeting.rs]: @MSNTCS Could you check this comment?

File: votingsystem/src/general_meeting.rs [hunk 15]
```diff
@@ -0,0 +1,248 @@
...
+        self.storage.replace(storage.unwrap_import().into_remote());
+    }
+}
+
+impl InitGenesis for Context {
+    fn begin_genesis(&mut self) {}
+
+    fn init_genesis(&mut self, config: &[u8]) {
+        let admin: Public = serde_cbor::from_slice(&config).unwrap();
+        self.storage_mut().set(admin.as_ref(), serde_cbor::to_vec(&admin).unwrap());
```
Comment (majecty) [inline votingsystem/src/general_meeting.rs:95]: Why did you use the admin's public key as the "key"?
How will use read the value?

File: votingsystem/src/general_meeting.rs [hunk 16]
```diff
@@ -0,0 +1,248 @@
...
+        }
+    }
+
+    fn check_transaction(&self, transaction: &Transaction) -> Result<(), coordinator::types::ErrorCode> {
+        let todo_fixthis: coordinator::types::ErrorCode = 3;
+        assert_eq!(transaction.tx_type(), "GeneralMeeting");
+        let tx: OwnTransaction = serde_cbor::from_slice(&transaction.body()).map_err(|_| todo_fixthis)?;
+        tx.verify().map_err(|_| todo_fixthis)?;
+
+        let admin_public_key = tx.signer_public;
```
Comment (majecty) [inline votingsystem/src/general_meeting.rs:125]: Shouldn't we check the admin_public key is really admin?

File: votingsystem/src/general_meeting.rs (hunk 3 above)
Comment (MSNTCS) [reply votingsystem/src/general_meeting.rs]: Yes, that is what I meant.

File: votingsystem/src/general_meeting.rs (hunk 4 above)
Comment (MSNTCS) [reply votingsystem/src/general_meeting.rs]: Yes that is my intention.

File: votingsystem/src/general_meeting.rs (hunk 3 above)
Comment (majecty) [reply votingsystem/src/general_meeting.rs]: If we serialize a GeneralMeeting struct, it will be serialized like this:

```json
{
  tallying_time: {
    time: 3
  }
}
```

I think the format below is better.

```json
{
  tallying_time: 3
}
```

What do you think about it?

Also we can do it by adding a line over `Timestamp` struct: `#[serde(transparent)] `
https://serde.rs/container-attrs.html#transparent

Comment (MSNTCS) [reply votingsystem/src/general_meeting.rs]: I agree with your point. I will use ``#[serde(transparent)]``. Thanks for mentioning.

File: votingsystem/src/general_meeting.rs (hunk 16 above)
Comment (MSNTCS) [reply votingsystem/src/general_meeting.rs:125]: @majecty, please read the next three lines.

File: votingsystem/src/general_meeting.rs [hunk 17]
```diff
@@ -0,0 +1,251 @@
...
+        self.storage.replace(storage.unwrap_import().into_remote());
+    }
+}
+
+impl InitGenesis for Context {
+    fn begin_genesis(&mut self) {}
+
+    fn init_genesis(&mut self, config: &[u8]) {
+        let admin: Public = serde_cbor::from_slice(&config).unwrap();
+        self.storage_mut().set("admin".as_bytes(), serde_cbor::to_vec(&admin).unwrap());
```
Comment (majecty) [inline votingsystem/src/general_meeting.rs:95]: Let's declare a constant variable for "admin"

File: votingsystem/src/voting.rs [hunk 18]
```diff
@@ -0,0 +1,270 @@
...
+    InvalidFormat,
+    InvalidAdmin,
+}
+
+pub enum Id {
+    VotePaperId(u8),
+    BallotId(u8),
+}
+#[derive(Debug, Serialize, Deserialize)]
+pub struct VoteId {
```
Comment (majecty) [inline votingsystem/src/voting.rs:44]: Let's use serde::transparent.

File: votingsystem/src/voting.rs [hunk 19]
```diff
@@ -0,0 +1,270 @@
...
+    InvalidMetadata,
+    InvalidSign,
+    InvalidFormat,
+    InvalidAdmin,
+}
+
+pub enum Id {
+    VotePaperId(u8),
+    BallotId(u8),
+}
```
Comment (majecty) [inline votingsystem/src/voting.rs:42]: Where will you use this `Id`?

File: votingsystem/src/general_meeting.rs [hunk 20]
```diff
@@ -0,0 +1,251 @@
...
+#[derive(Debug, Serialize, Deserialize)]
+#[serde(transparent)]
+pub struct TimeStamp {
+    time: u64,
+}
+
+#[derive(Debug, Serialize, Deserialize, Clone)]
+#[serde(transparent)]
+pub struct GeneralMeetingId {
+    pub id: u8,
```
Comment (majecty) [inline votingsystem/src/general_meeting.rs:148]: u8 is too small. How about U256?

File: votingsystem/src/voting.rs [hunk 21]
```diff
@@ -0,0 +1,270 @@
...
+#[derive(Debug, Serialize, Deserialize)]
+pub struct VoteId {
+    id: u8,
```
Comment (majecty) [inline votingsystem/src/voting.rs:45]: u8 is too small.

File: votingsystem/src/voting.rs [hunk 22]
```diff
@@ -0,0 +1,270 @@
...
+
+impl AsRef<StorageKeyRef> for VoteId {
+    fn as_ref(&self) -> &[u8] {
+        std::slice::from_ref(&self.id)
+    }
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+struct BallotId {
+    id: u8,
```
Comment (majecty) [inline votingsystem/src/voting.rs:58]: u8 is too small.

File: votingsystem/src/voting.rs [hunk 23]
```diff
@@ -0,0 +1,270 @@
...
+        std::slice::from_ref(&self.id)
+    }
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+struct VotePaper {
+    general_meeting_id: GeneralMeetingId,
+    agenda_number: u32,
+    authorization_token: u64,
+    used_in: VoteId,
```
Comment (majecty) [inline votingsystem/src/voting.rs:71]: Please use `Option<VoteId>`. There will be vote papers that are not used yet.

File: votingsystem/src/voting.rs [hunk 24]
```diff
@@ -0,0 +1,270 @@
...
+        if transaction.tx_type() != "Vote" {
+            return Err(ExecuteError::InvalidMetadata)
+        }
+        let tx: OwnTransaction =
+            serde_cbor::from_slice(&transaction.body()).map_err(|_| ExecuteError::InvalidFormat)?;
+        tx.verify().map_err(|_| ExecuteError::InvalidSign)?;
+
+        let mut meeting_id = &GeneralMeetingId {
+            id: 0,
+        };
```
Comment (majecty) [inline votingsystem/src/voting.rs:163]: It looks strange.
The meeting_id is set in the "CreateVotePaper" action and used in the "Vote" action.
The can't be executed at the same function.

File: votingsystem/src/voting.rs [hunk 25]
```diff
@@ -0,0 +1,270 @@
...
+        match tx.tx.action {
+            TxAction::CreateVotePaper(TxCreateVotePaper {
+                general_meeting_id,
+                agenda_number,
+                authorization_token,
+                voter_public_key,
+            }) => {
+                let vote_paper =
+                    VotePaper::new(general_meeting_id.clone(), agenda_number, authorization_token, voter_public_key);
+                self.storage_mut().set(vote_paper.used_in.as_ref(), serde_cbor::to_vec(&vote_paper).unwrap());
```
Comment (majecty) [inline votingsystem/src/voting.rs:173]: When we create a vote paper, the vote_paper_used_in will be empty because it is not used yet.

File: votingsystem/src/voting.rs [hunk 26]
```diff
@@ -0,0 +1,270 @@
...
+                meeting_id = &general_meeting_id;
+                Ok(Id::VotePaperId(vote_paper.used_in.id))
+            }
+            TxAction::Vote(Ballot {
+                choice,
+                voter_signature,
+            }) => {
+                let ballot = Vote::new(choice, voter_signature);
+                let id = &ballot.vote_paper_id;
+                self.storage_mut().set(id.as_ref(), serde_cbor::to_vec(&ballot).unwrap());
```
Comment (majecty) [inline votingsystem/src/voting.rs:183]: This will remove the vote paper information. Why did you remove the vote paper form the state?

File: votingsystem/src/voting.rs (hunk 18 above)
Comment (MSNTCS) [reply votingsystem/src/voting.rs:47]: Done

File: votingsystem/src/general_meeting.rs (hunk 17 above)
Comment (MSNTCS) [reply votingsystem/src/general_meeting.rs:96]: Cannot make the admin constant due to the fact that the config value is not constant.

File: votingsystem/src/voting.rs [hunk 27]
```diff
@@ -0,0 +1,280 @@
...
+            Ok(Default::default())
+        }
+    }
+
+    fn check_transaction(&self, transaction: &Transaction) -> Result<(), coordinator::types::ErrorCode> {
+        let todo_fixthis: coordinator::types::ErrorCode = 3;
+        assert_eq!(transaction.tx_type(), "Vote");
+        let tx: OwnTransaction = serde_cbor::from_slice(&transaction.body()).map_err(|_| todo_fixthis)?;
+        tx.verify().map_err(|_| todo_fixthis)?;
+        Ok(())
```
Comment (majecty) [inline votingsystem/src/voting.rs:230]: Please check these things.

TxAction::CreateVotePaper:

* general meeting is exist or not
* general meeting is finished or not
* voting time is finished or not
* agenda number exceeds the agenda count in the general meeting or not

TxAction::Vote

* vote paper is already used or not
* choice should be one of 4 choices
* voter signature is generated from voter public key or not

File: votingsystem/src/voting.rs [hunk 28]
```diff
@@ -0,0 +1,280 @@
...
+    }
+}
+
+impl Service for Context {}
+
+#[derive(Serialize, Deserialize, Debug)]
+pub struct TxCreateVotePaper {
+    general_meeting_id: GeneralMeetingId,
+    agenda_number: u32,
+    authorization_token: H256,
```
Comment (majecty) [inline votingsystem/src/voting.rs:138]: What is the `authorization_token`?

Comment (MSNTCS) [reply votingsystem/src/voting.rs:138]: It is a token that the person gets after authentication. We should check whether the person is authorized or not but because we do not save their tokens in the first place it needs a mechanism for it.

File: votingsystem/src/common.rs [hunk 29]
```diff
@@ -0,0 +1,71 @@
...
+    }
+}
+
+impl Default for NetworkId {
+    fn default() -> Self {
+        NetworkId([b't', b'c'])
+    }
+}
+
+pub trait Action: Serialize + std::fmt::Debug {}
```
[the remark this answers is not in the document]
Comment (MSNTCS) [reply votingsystem/src/common.rs:39]: @majecty if you remove `Serialization` here and replace it with `Desrilization`, you would face a `life-time` error.

[the remark this answers is not in the document]
Comment (majecty) [reply votingsystem/src/common.rs:39]: Thanks, I'll check it.

[the remark this answers is not in the document]
Comment (majecty) [reply votingsystem/src/common.rs:39]: Sorry, serialzation is required to get the hash of the transaction.
https://github.com/CodeChain-io/foundry/pull/452/files#diff-725c366433452466115179ac7d399f0fR67 Let's keep it.

File: votingsystem/src/voting.rs (hunk 28 above)
Comment (majecty) [reply votingsystem/src/voting.rs]: I think my explanation was not good.

We can think about it like this:
Actually, creating a vote paper (with the public key) means that a "person" who has a signature is a stakeholder.
When a voter sends vote transaction, we can verify the vote by checking the signature is created by the "private key".

File: votingsystem/src/lib.rs [hunk 30]
```diff
@@ -0,0 +1,29 @@
...
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+pub const prefix: &str = "VoteBox";
+
+pub mod common;
+pub mod general_meeting;
+pub mod voting;
+
+pub fn generate_key(meeting_id: &general_meeting::GeneralMeetingId) -> Vec<u8> {
```
Comment (majecty) [inline votingsystem/src/lib.rs:23]: `generate_key` generates a key for "VoteBox", but the name does not be related to the "VoteBox".
Please make the name contains the "VoteBox" information.

File: votingsystem/src/general_meeting.rs [hunk 31]
```diff
@@ -0,0 +1,382 @@
...
+                        vote_result.favor += number_of_shares;
+                    }
+                    if vote_choice == crate::voting::against {
+                        vote_result.against += number_of_shares;
+                    }
+                    if vote_choice == crate::voting::absention {
+                        vote_result.absention += number_of_shares;
+                    }
+                }
+                Ok(TxResult::Result(vote_result))
```
Comment (majecty) [inline votingsystem/src/general_meeting.rs:140]: Shouldn't we save the vote_result in `GeneralMeeting` struct?

File: votingsystem/src/general_meeting.rs [hunk 32]
```diff
@@ -0,0 +1,382 @@
...
+                ExecuteError::InvalidAdmin => Err(()),
+            }
+        } else {
+            Ok(Default::default())
+        }
+    }
+
+    fn check_transaction(&self, transaction: &Transaction) -> Result<(), coordinator::types::ErrorCode> {
+        let todo_fixthis: coordinator::types::ErrorCode = 3;
+        assert_eq!(transaction.tx_type(), "GeneralMeeting");
```
Comment (majecty) [inline votingsystem/src/general_meeting.rs:183]: Let's make a constant for "GeneralMeeting"

File: votingsystem/src/general_meeting.rs (hunk 31 above)
Comment (majecty) [inline votingsystem/src/general_meeting.rs:140]: Please check the block's timestamp is later than the `tallying_time`.

File: votingsystem/src/general_meeting.rs [hunk 33]
```diff
@@ -0,0 +1,382 @@
...
+                    let number_of_shares = self.vote().get_shares(&vote_id).unwrap();
+                    if vote_choice == crate::voting::favor {
+                        vote_result.favor += number_of_shares;
+                    }
+                    if vote_choice == crate::voting::against {
+                        vote_result.against += number_of_shares;
+                    }
+                    if vote_choice == crate::voting::absention {
+                        vote_result.absention += number_of_shares;
+                    }
```
Comment (majecty) [inline votingsystem/src/general_meeting.rs:138]: The result should be generated per agenda.

File: votingsystem/src/voting.rs [hunk 34]
```diff
@@ -0,0 +1,419 @@
...
+    CreateVotePaper(TxCreateVotePaper),
+    Vote(TxVote),
+}
+
+impl Action for TxAction {}
+pub type OwnTransaction = crate::common::SignedTransaction<TxAction>;
+
+impl Context {
+    fn excute_tx(&mut self, transaction: &Transaction) -> Result<H256, ExecuteError> {
+        if transaction.tx_type() != "Vote" {
```
Comment (majecty) [inline votingsystem/src/voting.rs:225]: Let's make a constant for the "Vote".

File: votingsystem/src/voting.rs [hunk 35]
```diff
@@ -0,0 +1,419 @@
...
+                general_meeting_id,
+                agenda_number,
+                number_of_shares,
+                voter_name: voter_name,
+                voter_public_key,
+            }) => {
+                let meeting: GeneralMeeting = self.meeting_mut().get_meeting(&general_meeting_id).unwrap();
+                if meeting.id != general_meeting_id {
+                    return Err(ExecuteError::InvalidMeeting)
+                }
```
Comment (majecty) [inline votingsystem/src/voting.rs:243]: This comparison is strange. How the meeting has a differnt id when we query the meeting with the id.

File: votingsystem/src/voting.rs [hunk 36]
```diff
@@ -0,0 +1,419 @@
...
+
+        match tx.tx.action {
+            TxAction::CreateVotePaper(TxCreateVotePaper {
+                general_meeting_id,
+                agenda_number,
+                number_of_shares,
+                voter_name: voter_name,
+                voter_public_key,
+            }) => {
+                let meeting: GeneralMeeting = self.meeting_mut().get_meeting(&general_meeting_id).unwrap();
```
Comment (majecty) [inline votingsystem/src/voting.rs:240]: Please do not use `unwrap` here.
It will make Voting module crash when the CreateVotePaper transaction has an invalid general_meeting_id.

File: votingsystem/src/voting.rs [hunk 37]
```diff
@@ -0,0 +1,419 @@
...
+                if meeting.id != general_meeting_id {
+                    return Err(ExecuteError::InvalidMeeting)
+                }
+                if meeting.tallying_time.time < self.block_header.as_ref().unwrap().timestamp() {
+                    return Err(ExecuteError::InvalidMeetingTime)
```
Comment (majecty) [inline votingsystem/src/voting.rs:245]: The error name looks not good.
When I read the error name I thought that the time in the general meeting is invalid.

Actually, the `tallying_time` in the `GeneralMeeting` is valid. What is incorrect is that the vote paper is created after tallying. Please change the error name.

File: votingsystem/src/voting.rs [hunk 38]
```diff
@@ -0,0 +1,419 @@
...
+                }
+                if meeting.number_of_agendas < agenda_number {
+                    return Err(ExecuteError::InvalidNumberofAgendas)
+                }
```
Comment (majecty) [inline votingsystem/src/voting.rs:249]: Does the agenda number start from 0? or 1?

If it starts from 0, this condition is wrong. The condition should be `meeting.number_of_agendas <= agenda_number`.
If it starts from 1, this condtion is wrong. The condition should be `0 < agenda_number && agenda_number <= meeting.number_of_agendas`.

File: votingsystem/src/voting.rs [hunk 39]
```diff
@@ -0,0 +1,419 @@
...
+                if meeting.voting_end_time.time > self.block_header.as_ref().unwrap().timestamp() {
+                    return Err(ExecuteError::InvalidVotingTime)
```
Comment (majecty) [inline votingsystem/src/voting.rs:251]: Please the name of the error.
`VotingTime` is correct. The problem is that `CreateVotePaper` transaction is included in a block too late.

File: votingsystem/src/voting.rs [hunk 40]
```diff
@@ -0,0 +1,419 @@
...
+            TxAction::Vote(TxVote {
+                vote_paper_id,
+                choice,
+                voter_signature,
+            }) => {
+                if choice != favor || choice != against || choice != absention {
+                    return Err(ExecuteError::InvalidChoice)
+                }
+                let mut vote_paper: VotePaper =
+                    serde_cbor::from_slice(&self.storage().get(vote_paper_id.0.as_ref()).unwrap()).unwrap();
```
Comment (majecty) [inline votingsystem/src/voting.rs:273]: What happens if there is no VotePaper?

File: votingsystem/src/voting.rs [hunk 41]
```diff
@@ -0,0 +1,419 @@
...
+                let corresponding_paper_id = vote_paper.vote_paper_id;
+                let vote = Vote::new(vote_paper_id, choice, voter_signature);
+                let vote_id = vote.clone().vote_id;
+                vote_paper.set_used_in(vote_id.clone());
```
Comment (majecty) [inline votingsystem/src/voting.rs:277]: Please check the vote_paper.used_in is already set or not.

File: votingsystem/src/voting.rs [hunk 42]
```diff
@@ -0,0 +1,419 @@
...
+                    voter_public_key,
+                );
+                self.storage_mut().set(vote_paper.vote_paper_id.0.as_ref(), serde_cbor::to_vec(&vote_paper).unwrap());
+                Ok(vote_paper.vote_paper_id)
+            }
+
+            TxAction::Vote(TxVote {
+                vote_paper_id,
+                choice,
+                voter_signature,
```
Comment (majecty) [inline votingsystem/src/voting.rs:267]: Please check the voter_signature is generated from the voter private key.
Since we have voter public key, we can verify it.

File: votingsystem/src/voting.rs [hunk 43]
```diff
@@ -0,0 +1,419 @@
...
+                agenda_number,
+                voter_name,
+                number_of_shares,
+                voter_public_key,
+            }) => {
+                let meeting: GeneralMeeting =
+                    serde_cbor::from_slice(&self.storage().get(general_meeting_id.as_ref()).unwrap()).unwrap();
+                if meeting.id != general_meeting_id {
+                    return Err(ExecuteError::InvalidMeeting).map_err(|_| todo_fixthis)
+                }
```
Comment (majecty) [inline votingsystem/src/voting.rs:337]: When the returned meeting.id != general_meeting_id?

File: votingsystem/src/voting.rs [hunk 44]
```diff
@@ -0,0 +1,419 @@
...
+        match tx.tx.action {
+            TxAction::CreateVotePaper(TxCreateVotePaper {
+                general_meeting_id,
+                agenda_number,
+                voter_name,
+                number_of_shares,
+                voter_public_key,
+            }) => {
+                let meeting: GeneralMeeting =
+                    serde_cbor::from_slice(&self.storage().get(general_meeting_id.as_ref()).unwrap()).unwrap();
```
Comment (majecty) [inline votingsystem/src/voting.rs:334]: Please handle the case when the GeneralMeeting does not exist.

File: votingsystem/src/voting.rs [hunk 45]
```diff
@@ -0,0 +1,419 @@
...
+                if meeting.id != general_meeting_id {
+                    return Err(ExecuteError::InvalidMeeting).map_err(|_| todo_fixthis)
+                }
+                if meeting.number_of_agendas < agenda_number {
```
Comment (majecty) [inline votingsystem/src/voting.rs:338]: Does the agenda number start from 0? or 1?

If it starts from 0, this condition is wrong. The condition should be `meeting.number_of_agendas <= agenda_number`.
If it starts from 1, this condtion is wrong. The condition should be `0 < agenda_number && agenda_number <= meeting.number_of_agendas`.

File: votingsystem/src/voting.rs (hunk 41 above)
Comment (MSNTCS) [reply votingsystem/src/voting.rs:277]: @majecty set_used_in function checks this.

Comment (majecty) [reply votingsystem/src/voting.rs:277]: You should check the returned result of the `set_used_in` function.
If you did not check the returned value, actually you didn't check the property.

File: votingsystem/src/voting.rs (hunk 38 above)
Comment (MSNTCS) [reply votingsystem/src/voting.rs:249]: @majecty it is started from 0. if the given number of agendas is equal to the actual number of agendas; there should not be any error the error happens when the given agenda is greater than the actual one.

File: votingsystem/src/voting.rs (hunk 42 above)
Comment (MSNTCS) [reply votingsystem/src/voting.rs:267]: @MSNTCS The verification happens when `execute_tx` is called.

File: voting/src/general_meeting.rs [hunk 46]
```diff
@@ -0,0 +1,448 @@
...
+                    serde_cbor::from_slice(&transaction.body()).map_err(|_| ExecuteError::InvalidFormat)?;
+
+                let meeting_id = tx.action.meeting_id;
+                let key = crate::generate_voting_box_key(&meeting_id);
+                let vote_box: VoteBox = serde_cbor::from_slice(&self.storage().get(key.as_slice()).unwrap()).unwrap();
+                let mut vote_result = VoteResult::new();
+
+                let mut meeting: GeneralMeeting =
+                    serde_cbor::from_slice(&self.storage().get(meeting_id.as_ref()).unwrap())
+                        .map_err(|_| ExecuteError::InvalidMetadata)?;
```
Comment (majecty) [inline voting/src/general_meeting.rs:147]: What is the the `Metadata`?

Comment (majecty) [inline voting/src/general_meeting.rs:147]: Could you split this line into multiple lines? Since there are multiple `unwrap`s, it is hard to read the role of unwraps.

I also changed the error handling. We should handle the case when the general meeting does not exist.
We don't need to handle the case when the serde deserialization fails. Since the state is only readable and writable by this module we can trust the deserialization.

```suggestion
                let mut meeting: GeneralMeeting = {
                    let bytes =
                        self.storage().get(meeting_id.as_ref()).ok_or_else(|| ExecuteError::GeneralMeetingNotFound)?;

                    serde_cbor::from_slice(&bytes)
                    .expect("General meeting is serialized by this code")
                };
```

File: voting/Cargo.toml [hunk 47]
```diff
@@ -0,0 +1,21 @@
+[package]
+name = "codechain-votingsystem"
+version = "0.1.0"
+authors = ["CodeChain Team <hi@codechain.io>"]
+edition = "2018"
+
+[dependencies]
+ctypes = { package = "codechain-types", path = "../types" }
+coordinator = { path = "../coordinator" }
+ccrypto = { package = "codechain-crypto", git = "https://github.com/CodeChain-io/rust-codechain-crypto.git", version = "0.2" }
```
[the remark this answers is not in the document]
Comment (MSNTCS) [reply voting/Cargo.toml:10]: > If you don't know an overview of how dependencies work in general, I can explain it to you briefly.

I know it, thanks

File: voting/src/common.rs [hunk 48]
```diff
@@ -0,0 +1,71 @@
...
+            Ok(())
+        } else {
+            Err(())
+        }
+    }
+}
+
+#[derive(Serialize, Deserialize, Debug)]
+pub struct UserTransaction<T: Action> {
+    pub seq: u64,
```
[the remark this answers is not in the document]
Comment (MSNTCS) [reply voting/src/common.rs:61]: Now that I am thinking about it, we do not need it.

**Comment by MSNTCS** [comment]
This PR is split into several small PRs. The list of PRs is as follows: 
 #459 ~ #464 and #466 ~ #472.
